### PR TITLE
basic support for ledger rebuilds when editing submissions

### PR DESCRIPTION
**Pull request by czue**
http://manage.dimagi.com/default.asp?181166

@dannyroberts 

wanted to run tests, but also think we could just merge this as-is if they pass.

probably easiest to review commit-wise

cc @benrudolph

### CODE REVIEW

File: corehq/apps/commtrack/tests/util.py [hunk 1]
```diff
@@ -202,6 +204,27 @@ def extract_balance_xml(xml_payload):
...
+    <entry id="{product_id}" quantity="{quantity}" />
+</balance>""".format(
+        case_id=case_id, product_id=product_id, quantity=quantity, date=date_string, section_id=section_id
+    ).strip()
+
+
+def get_single_transfer_block(src_id, dest_id, product_id, quantity, date_string=None, section_id='stock'):
+    date_string = date_string or json_format_date(datetime.utcnow())
+    return """
+<transfer xmlns="http://commcarehq.org/ledger/v1" src="{src_id}" dest="{dest_id}" date="{date}" section-id="{section_id}">
```
Comment (dimagimon) [inline corehq/apps/commtrack/tests/util.py]: E501 line too long (122 > 115 characters)

**Comment by czue** [comment]
just tacked on a bonus fix for this interrupt bug: http://manage.dimagi.com/default.asp?181083

File: corehq/apps/commtrack/processing.py [hunk 2]
```diff
@@ -2,12 +2,13 @@
 import logging
 from django.db import transaction
 from django.utils.translation import ugettext as _
+from casexml.apps.case.cleanup import deprecate_actions_for_form
```
Comment (dimagimon) [inline corehq/apps/commtrack/processing.py]: F401 'deprecate_actions_for_form' imported but unused

File: corehq/apps/commtrack/parsing.py [hunk 3]
```diff
@@ -51,14 +51,15 @@
...
 from casexml.apps.stock import const as stockconst
 from casexml.apps.stock.const import COMMTRACK_REPORT_XMLNS
 from corehq.apps.commtrack import const
 from corehq.apps.commtrack.exceptions import InvalidDate
-from corehq.apps.commtrack.models import StockReportHelper, \
-    StockTransactionHelper
+from corehq.apps.commtrack.models import StockReportHelper, StockTransactionHelper
 from couchforms.models import XFormInstance
-from couchforms.util import adjust_datetimes
+from couchforms.util import adjust_datetimes, is_deprecation
```
Comment (dimagimon) [inline corehq/apps/commtrack/parsing.py]: F401 'is_deprecation' imported but unused

**Comment by czue** [comment]
actually managed to get the most basic test case for this working. given that this is significantly better than the previous version I think I'm comfortable with this getting merged after reveiw

File: corehq/apps/commtrack/processing.py [hunk 4]
```diff
@@ -43,10 +43,23 @@ def commit(self):
...
+def delete_models_for_stock_report(domain, stock_report_helper):
+    """
+    Delete all stock reports and stock transaction models associated with the helper from the database.
+    """
+    assert stock_report_helper.domain == domain
+    StockReport.objects.filter(domain=domain, form_id=stock_report_helper.form_id).delete()
+
+
+@transaction.atomic
+@transaction.atomic
```
Comment (dannyroberts) [inline corehq/apps/commtrack/processing.py]: Looks like there are two of these

Comment (czue) [inline corehq/apps/commtrack/processing.py]: whoops, thanks

**Comment by czue** [comment]
also as you may have noticed - this doesn't actually rebuild the stock state from the transactions yet - though i'm tempted to advocate for merge since i'm not sure when I'll add/test that last bit

**Comment by dannyroberts** [comment]
@czue this looks pretty awesome. I was starting to hit a wall with refactoring fatigue on this code, glad you were able to pick up where I left off and be so productive so quickly!

**Comment by dannyroberts** [comment]
Taking a closer look now...

File: corehq/apps/commtrack/processing.py [hunk 5]
```diff
@@ -43,10 +43,23 @@ def commit(self):
 
         # create the django models
         for stock_report_helper in self.stock_report_helpers:
-            create_models_for_stock_report(self.domain, stock_report_helper)
+            if stock_report_helper.deprecated:
+                delete_models_for_stock_report(self.domain, stock_report_helper)
+            else:
+                create_models_for_stock_report(self.domain, stock_report_helper)
```
Comment (dannyroberts) [inline corehq/apps/commtrack/processing.py]: These lines look like the crux of the change, and everything else is mostly in service to this change—is that right?

File: corehq/apps/commtrack/processing.py [hunk 6]
```diff
@@ -151,48 +172,70 @@ def get_stock_actions(xform):
...
+        else:
+            case_action = CommCareCaseAction.from_parsed_action(
+                submit_time, user_id, xform, AbstractAction(CASE_ACTION_COMMTRACK)
+            )
+            case_action_intents.append(CaseActionIntent(
+                case_id=case_id, form_id=xform._id, is_deprecation=False, action=case_action
+            ))
 
-    return stock_report_helpers, case_actions
+    return StockFormActions(stock_report_helpers, case_action_intents)
```
Comment (dannyroberts) [inline corehq/apps/commtrack/processing.py]: Very nice change

File: corehq/apps/commtrack/processing.py (hunk 5 above)
Comment (czue) [inline corehq/apps/commtrack/processing.py]: this and the case actions changes which I'll highlight now.

File: corehq/apps/commtrack/processing.py [hunk 7]
```diff
@@ -151,48 +171,70 @@ def get_stock_actions(xform):
...
         if case is None:
             raise IllegalCaseId(
                 _('Ledger transaction references invalid Case ID "{}"')
                 .format(case_id))
 
-        # hack: clear the sync log id so this modification always counts
-        # since consumption data could change server-side
-        case_action.sync_log_id = ''
-        case.actions.append(case_action)
+        if action_intent.is_deprecation:
```
Comment (czue) [inline corehq/apps/commtrack/processing.py]: (aforementioned case actions changes)

**Comment by czue** [comment]
not sure if will be useful but I [outlined the changes i was going to make](https://docs.google.com/document/d/10sbL9zvgazP66EOLHEekAXRtTrU1epIGD29mWPI7pOc/edit#heading=h.p5hz5hg9lywi) before i made them in case helpful (it ended up working out surprisingly close to the plan).

and yeah - all the prior refactors helped a ton! it was very nice to have access to this code in the approximate places I would have expected it to be and not have to wrangle with all sorts of argument passing up and down crazy call chains...

File: corehq/apps/commtrack/processing.py (hunk 7 above)
Comment (dannyroberts) [inline corehq/apps/commtrack/processing.py]: Nice

**Comment by dannyroberts** [comment]
Yeah, looks good. I like that the code now does something significantly more complex but if anything is more readable than it started! :star:

And so I understand your earlier comment—and I think this is what I'm seeing in the code—is the right now this code deletes the deprecated form's transactions and applies the new one's on top of the resulting state, but doesn't go back and try to insert them into the right order and rebuild or anything. That sounds good to me—and agree the last part is a little tricky and worth separating out from this.

**Comment by czue** [comment]
> right now this code deletes the deprecated form's transactions and applies the new one's on top of the resulting state, but doesn't go back and try to insert them into the right order and rebuild or anything

exactly. thanks for quick turnaround!

**Pull request merged**